cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,492 @@
+                        default/None: nothing displayed
+                - time_limit: stop after this many seconds (default: None)
+                - solution_limit: stop after this many solutions (default: None)
+                - any other keyword argument
+
+            Returns: number of solutions found
+        """
+
+        assert not self.has_objective # TODO: is this true?
+
+        if not self.solver_is_initialized:
+            self.xct_solver.init([],[])
+            self.solver_is_initialized=True
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            print(kw,val)
+            print(type(self.xct_solver))
+            self.xct_solver.setOption(kw,str(val))
+
+        solsfound = 0
+        while solution_limit == None or solsfound < solution_limit:
+            # call the solver, with parameters
+            my_status = self.xct_solver.runFull(self.has_objective,time_limit if time_limit is not None else 0)
+            assert my_status in [0,1,2,3], "Unexpected status code for Exact."
+            if my_status == 0: # found unsatisfiability
+                break
+            elif my_status == 1: # found solution, but not optimality proven
+                assert self.xct_solver.hasSolution()
+                solsfound += 1
+                self.getSolAndObj()
+                self.xct_solver.invalidateLastSol()
+            elif my_status == 2: # found inconsistency
+                pass # TODO
+            elif my_status == 3: # found timeout
+                pass # TODO
+
+        return solsfound
+
+
+    def solver_var(self, cpm_var, encoding="onehot"):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        if is_num(cpm_var):  # shortcut, eases posting constraints
+            return cpm_var
+
+        # variables to be translated should be positive
+        assert(not isinstance(cpm_var, NegBoolView))
+
+        # return it if it already exists
+        if cpm_var in self._varmap:
+            return self._varmap[cpm_var]
+
+        # create if it does not exist
+        revar = str(cpm_var)
+        if isinstance(cpm_var, _BoolVarImpl):
+            self.xct_solver.addVariable(revar,0,1)
+        elif isinstance(cpm_var, _IntVarImpl):
+            lb = cpm_var.lb
+            ub = cpm_var.ub
+            if max(abs(lb),abs(ub)) > 1e18:
+                # larger than 64 bit should be passed by string
+                self.xct_solver.addVariable(revar,str(lb), str(ub), encoding)
+            else:
+                self.xct_solver.addVariable(revar,lb,ub, encoding)
+        else:
+            raise NotImplementedError("Not a known var {}".format(cpm_var))
+        self._varmap[cpm_var] = revar
+        return revar
+    
+
+    def solver_var_no_num(self, cpm_var, encoding="onehot"):
```

Comment:
can you document the purpose of this function?

Reply:
Done. The idea is to just have a private wrapper that checks whether the invariant holds that whatever we request is not a number. This should be the case, otherwise things will break internally. The external `solver_var` can handle numbers (e.g., for `DirectConstraint`).

If you're concerned about efficiency, I can optimize the wrapper a bit.

Reply:
removed the wrapper in the end.